hack/verify-typecheck.sh

```diff
@@ -26,6 +26,17 @@ source "${KUBE_ROOT}/hack/lib/init.sh"
 cd "${KUBE_ROOT}"
 
 kube::golang::setup_env
+kube::util::require-jq
+
+if [[ $# == 0 ]]; then
+  # Doing it this way is MUCH faster than simply saying "all", and there doesn't
+  # seem to be a simpler way to express "this whole workspace".
+  packages=()
+  kube::util::read-array packages < <(
```

Comment:
@thockin this seems like something we should move to a utility in hack/lib/golang.sh ?

Reply:
Arrays and functions don't always play nice together - you can't pass an array as an argument, for example.  I suppose we could do wrap it, but it's only used in a couple places

Reply:
> Arrays and functions don't always play nice together - you can't pass an array as an argument, for example. I suppose we could do wrap it, but it's only used in a couple places

I don't see why calling this snippet would be any different? `kube::golang::read-packages packages` ?

> [...] but it's only used in a couple places

Yeah, but this seems likely to need updating at some point, it's a bit of a hack?

Reply:
I don't think it's a hack.  Go has a small number of syntaxes which indicate packages.  I don't know if there is a spec, per se, but other than the ones we all know ("example.com/path, ./path, and /... suffixes) there's a magic "all" pattern.  Last I asked there was no plan to add a magic pattern for "everything in this workspace", so  this is the way.

If you want to PR a function like `kube::golang::workspace_all()` I won't say no :)

Reply:
> I don't think it's a hack. Go has a small number of syntaxes which indicate packages. I don't know if there is a spec, per se, but other than the ones we all know ("example.com/path, ./path, and /... suffixes) there's a magic "all" pattern. Last I asked there was no plan to add a magic pattern for "everything in this workspace", so this is the way.

I meant that it feels slightly magical to jq `go work edit -json`, which isn't exactly a well-known data format (seems to be documented in https://go.dev/ref/mod#go-work-edit) and it seems likely this will get copy pasted a lot.


I started iterating here: https://github.com/kubernetes/kubernetes/pull/125964 and already found at two different implementations of this in tree already and three copy-pastes of this variation (including comments etc).

We can use `go list` to avoid pulling in jq (and checking for jq) at the expense of absolute paths.